[IMP] mail: adds closeThread prop and button

thjo-odoo opened the pull request:
This commit adds a new optional prop to the Message and Thread Components: `closeThread`. This prop is a function that is launched to close a comment thread in a knowledge.article via the first message of the thread.

We also added a button inside the actionBox in the Message Component amd it is only rendered if we are on the first message of the thread, the one that we cannot edit nor reply to, and if the prop exists.

ENT PR: odoo/enterprise#41951

task-3317056

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/127380).

awa-odoo commented:
Hello @thjo-odoo 

Do these new prop absolutely HAVE to be in the base components / templates of the mail module?

Or can we create a "KnowledgeCommentsThread" component that extends Thread in the knowledge module and adds these props, and specify another template that is itself an extension of the base template with our minor tweaks?
Then in knowledge you would use `<KnowledgeCommentsThread.../>` instead of Thread?

It would be best to keep those localized to our use case if possible.

Cheers!

alexkuhn commented:
Personally I don't mind a few new props in `mail`, patches can sometimes make code harder to maintain.

At a glance, only the props `openThread` and `closeThread` seem bad design. If the intent in `Message` component is to disable actions, you should pass `hasActions="false"`. These functions are also heavily coupled with the new thread actions in `knowledge`, it would be best to define the whole behaviour in the actions themselves, rather having the thread actions call a function on the `Thread` component.

thjo-odoo commented:
@alexkuhn Hello I changed the implementation for the 2 problematic props and changed them to only decide whether the button needs to be displayed or not => the action is now done within our message_actions inside `Knowledge`. Do you want me to change anything else in there ?

alexkuhn commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -84,6 +86,11 @@ export class Message extends Component {
         "thread?",
         "messageSearch?",
         "className?",
+        "openThread?",
+        "closeThread?",
```
Can remove these props, as they are now passed with `env` in knowledge

alexkuhn commented on addons/mail/static/src/core/common/message.xml:
```diff
@@ -121,7 +122,7 @@
 
 <t t-name="mail.Message.actions">
-    <div t-if="props.hasActions and message.hasActions" class="o-mail-Message-actions"
+    <div t-if="props.hasActions and message.hasActions and ((!env.closeThread and !env.openThread) or !state.isEditing)" class="o-mail-Message-actions"
```
knowledge can set `hasActions=False` to the message. You can introduce a new prop `messageActions=true/false` on component `Thread` to allow this customisation from `Thread` component which is what's used by knowledge

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -39,6 +39,10 @@
                         thread="props.thread"
                         messageEdition="props.messageEdition"
+                        avatarPlaceholder="props.avatarPlaceholder"
+                        closeThread="!props.threadResolved and msg_first"
+                        openThread="props.threadResolved and msg_first"
```
`openThread` and `closeThread` are passed with `env` in knowledge so they must be removed

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -3,7 +3,7 @@
 <t t-name="mail.Thread">
     <t t-if="env.inChatter" t-call="mail.Thread.jumpPresent"/> <!-- chatter has its own scrollable, this ensures proper sticky showing -->
-    <div class="o-mail-Thread position-relative flex-grow-1 d-flex flex-column overflow-auto" t-att-class="{ 'pb-4': !state.showJumpPresent }" t-ref="messages" tabindex="-1">
+    <div class="o-mail-Thread position-relative flex-grow-1 d-flex flex-column overflow-auto" t-att-class="{ 'pb-4': !state.showJumpPresent and !props.closeThread }" t-ref="messages" tabindex="-1">
```
If you want to disable the "Jump to present", make a prop that is named and specific to that intention

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -53,7 +57,7 @@
         </t>
         <t t-else="">
-            <div class="o-mail-Thread-empty d-flex flex-column align-items-center justify-content-center p-4 text-muted fst-italic h-100">
+            <div class="o-mail-Thread-empty d-flex flex-column align-items-center justify-content-center text-muted fst-italic h-100" t-att-class="{'p-4': !props.closeThread}">
```
Again if you want to customise showing/hiding empty text, make a dedicate prop for that

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -53,7 +57,7 @@
         </t>
         <t t-else="">
-            <div class="o-mail-Thread-empty d-flex flex-column align-items-center justify-content-center p-4 text-muted fst-italic h-100">
+            <div class="o-mail-Thread-empty d-flex flex-column align-items-center justify-content-center text-muted fst-italic h-100" t-att-class="{'p-4': !props.showEmptyThreadMessage}">
```
Note that "Thread" is noise in this prop name, as this is a prop on component `Thread`

alexkuhn replied:
And I think the condition must be inverted? Keep `p-4` with showing empty message

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -3,7 +3,7 @@
 <t t-name="mail.Thread">
     <t t-if="env.inChatter" t-call="mail.Thread.jumpPresent"/> <!-- chatter has its own scrollable, this ensures proper sticky showing -->
-    <div class="o-mail-Thread position-relative flex-grow-1 d-flex flex-column overflow-auto" t-att-class="{ 'pb-4': !state.showJumpPresent }" t-ref="messages" tabindex="-1">
+    <div class="o-mail-Thread position-relative flex-grow-1 d-flex flex-column overflow-auto" t-att-class="{ 'pb-4': !state.showJumpPresent and !props.showJumpPresent }" t-ref="messages" tabindex="-1">
```
`and props.showJumpPresent` (not `!`).

Nitpicking but it feels easier to read with `props.showJumpPresent` before `state.showJumpPresent`. The former is a configuration, the other is dynamic state, it feels more natural config before dynamic state in the conditions

alexkuhn commented on addons/mail/static/src/core/common/thread.xml:
```diff
@@ -76,7 +80,7 @@
                     </t>
                     </t>
-                    <t t-else="">
+                    <t t-elif="!props.showEmptyMessages">
```
Again that's likely the opposite: show it when prop is truthy

alexkuhn commented on addons/mail/static/src/core/common/thread.js:
```diff
@@ -32,18 +32,30 @@ export const PRESENT_THRESHOLD = 2500;
         "messageToReplyTo?",
         "order?",
         "scrollRef?",
+        "threadResolved?",
+        "showEmptyThreadMessage?",
+        "showJumpPresent?",
+        "messageActions?",
     ];
     static defaultProps = {
         isInChatWindow: false,
         jumpPresent: 0,
         order: "asc",
+        showDates: true,
+        avatarPlaceholder: false,
+        threadResolved: false,
+        showEmptyThreadMessage: true,
```
Can remove `threadResolved` (unused)

alexkuhn commented on addons/mail/static/src/core/common/message.js:
```diff
@@ -84,6 +86,9 @@ export class Message extends Component {
         "className?",
+        "avatarPlaceholder?",
+        "showDates?",
+        "isFirstMessage?"
```
I haven't tested in practice whether this feels good or not, just a random reflexion by just looking at the code.
You can ignore it if you think it works better that way.

Looks like this props is to limit the open/close thread to the 1st message in list of a knowledge article. I don't think this is intuitive UI/UX when only 1 message has it and its the 1st in list.

In Discuss we have "thread actions", which are available in both the Discuss App in a chat window:
![Screenshot 2023-10-20 at 16 50 52](https://github.com/odoo/odoo/assets/6569390/6206d83a-ac5b-467e-882f-ba871249b968)
![Screenshot 2023-10-20 at 16 50 59](https://github.com/odoo/odoo/assets/6569390/6c28afc9-894e-4f46-8a32-4844d11a103e)

As the name suggestion, it's like message actions but on a thread. These thread actions are custom displayed by the user of `Thread` component, and in discuss this is handled by `Discuss` and `ChatWindow` components. Probably the container of `Thread` component in knowledge could display the thread actions somewhere? Doesn't mean all actions are visible, you could decide to show only some actions and up to knowledge to choose how to display it, but I think it's best to make it more readily available on `Thread` container rather than hidden in a `Message` action.

thjo-odoo replied:
This is a good idea, I need to discuss this with LNA to see if he would be ok with this in the v2. But I'll keep it like this for now as it was the design that was specced at the start of this project.

tde-banana-odoo reviewed (approved):
If Discuss agrees, I agree :D .

robodoo commented:
@thjo-odoo @tde-banana-odoo linked pull request(s) odoo/enterprise#41951 not ready. Linked PRs are not staged until all of them are ready.
